File: ios/Sources/WalletExtensionPlugin/WalletSessionStore.swift
import Foundation

final class WalletSessionStore {
    private let storage = KeychainStorage(
        service: "com.3land.capacitor-wallet-extension.session",
        synchronizable: false
    )
    private let account = "connected-wallet"
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private(set) var currentSession: ConnectedWalletSession?

    init() {
        currentSession = try? load()
    }

    func save(_ session: ConnectedWalletSession) throws {
        let data = try encoder.encode(session)
        try storage.save(data, account: account)
        currentSession = session
    }

    func clear() throws {
        try storage.delete(account: account)
        currentSession = nil
    }

    private func load() throws -> ConnectedWalletSession? {
        guard let data = try storage.load(account: account) else {
            return nil
        }

        return try decoder.decode(ConnectedWalletSession.self, from: data)
    }
}

File: android/src/main/java/com/threeland/capacitor/walletextension/WalletSessionStore.java
package com.threeland.capacitor.walletextension;

import android.content.Context;

final class WalletSessionStore {
    private static final String SESSION_STORAGE_KEY = "connected-wallet-session";

    private final SecureStorage secureStorage;

    WalletSessionStore(Context context) {
        secureStorage = new SecureStorage(
            context,
            "com.3land.capacitor-wallet-extension.session.master"
        );
    }

    String load() {
        try {
            return secureStorage.loadString(SESSION_STORAGE_KEY);
        } catch (Exception ignored) {
            secureStorage.delete(SESSION_STORAGE_KEY);
            return null;
        }
    }

    void save(String session) throws Exception {
        secureStorage.saveString(SESSION_STORAGE_KEY, session);
    }

    void clear() {
        secureStorage.delete(SESSION_STORAGE_KEY);
    }
}
